packages/cli/src/ui/utils/textOutput.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Internal state for the singleton
+let lastChar = '\n';
+
+/**
+ * A singleton utility to manage writing text to stdout, ensuring that newlines
+ * are handled consistently and robustly across the application.
+ */
+export const textOutput = {
+  /**
+   * Writes a string to stdout.
+   * @param str The string to write.
+   */
+  write(str: string): void {
+    if (str.length === 0) {
+      return;
+    }
+    process.stdout.write(str);
+    lastChar = str.charAt(str.length - 1);
+  },
+
+  /**
+   * Writes a string to stdout, ensuring it starts on a new line.
+   * If the previous output did not end with a newline, one will be added.
+   * This prevents adding extra blank lines if a newline already exists.
+   * If no string is provided, it just ensures the output ends with a newline.
+   * @param str The optional string to write.
+   */
+  writeOnNewLine(str?: string): void {
+    if (lastChar !== '\n') {
+      this.write('\n');
+    }
+    if (str !== undefined) {
+      this.write(str);
+    }
+  },
+
+  /**
```

Comment:
if this wasn't a singleton you wouldn't have this issue testing. I would make this an instance so you can test easily. Given you are only using this in nonInteractiveCli there is no need to have it be a singleton. you can always pass it around or change it to a singleton later if that is needed.